binary_search_networks/search.py

```diff
@@ -0,0 +1,206 @@
+
+	# If there are not previously recorded slopes, model the probability if 1/yi because we expect a low probability of a nonzero slope but a high probability of a nonzero slope
+	if len(mx) == 0:
+		likelihood = norm(1/yi, sigma).pdf(1/yi)
+
+	# If there is only one previously recorded slope, model the probability simply with the first recorded slope and a general sigma
+	elif len(mx) == 1:
+		likelihood = norm(my[0], sigma).pdf(yi)
+
+	# If there are more than one recorded slopes, then model the probability using the linear regression relationship... this may be adapted to be a polynomial if it does not fit it well
+	else:
+		x = np.array(mx).reshape((-1, 1))
+		y = np.array(my)
+
+		model = LinearRegression().fit(x, y)
+		my_pred = model.predict(mx)
+		sigma = np.std(my_pred, my)
+
+		y_pred = model.predict(xi)
+
+		likelihood = norm(y_pred, sigma).pdf(yi)
+
+	if yi < 0:
+		likelihood *= -1
+
+	pior = delta / (gamma2 - gamma1)
+
+	return likelihood * pior
+
+
+def binary_search(**args):
+	'''
+	Purpose: find the maximum accuracy using binary search
+	Returns:
+	... mid: the found maximum accuracy
+	'''
+
+	# Gamma1 & gamma2 keep track of the current recorded upper and lower bounds to the number of units.
+	gamma1 = 1
+	gamma2 = args['n']
+
+	delta = args['delta']
+
+	# This is a threshold for when there is sufficient evidence that there is a maximum between ni & nj
+	posterior_alpha = args['posterior_alpha']
+
+	m1 = []
+	m2 = []
+
+	mid1 = []
+	mid2 = []
+	print(gamma1)
+	print(gamma2)
+	while gamma1 <= gamma2:
+		mid = (gamma1 + gamma2)//2
+		args['ni'] = mid - delta//2
+		args['nj'] = mid + delta//2
+
+		mi = get_slope(**args)
+
+		# When we are on the left side of the maximum
+		if mi > 0:
+			m1.append(mi)
+			mid1.append(mid)
+			args['ni'] = mid
+			# Get posterior probability (and if its sufficient, check the secant line on the respective side)
+			if get_posterior_prob(gamma1, gamma2, mid1, m1, delta) > posterior_alpha:
+				if get_slope(**args) < mi: # check if the slopes in between?
+					print("Maximum accuracy found at index {}".format(mid))
+					# TODO: decide if delta is sufficiently small than we can stop the search
+					return mid
+				# if delta is large (~50) such that the posterior begins to increase, decrease delta
+				else:
+					if delta > 3:
+						delta /= 2
+					elif delta < 6:
+						delta = 3
+			else:
+				gamma1 = mid # + 1?
+		# When we are on the right side of the maximum
+		else:
+			m2.append(mi)
+			mid2.append(mid)
+			args['nj'] = mid
+
+			# Get posterior probability (and if its sufficient, check the secant line on the respective side)
+			if get_posterior_prob(gamma1, gamma2, mid2, m2, delta) > posterior_alpha:
+				if get_slope(**args) > mi:
+					print("Maximum accuracy found at index {}".format(mid))
+					# TODO: decide if delta is sufficiently small than we can stop the search
+					return mid
+				# if delta is large (~50) such that the posterior begins to increase, decrease delta
+				else:
```

Comment:
similar to [adaptive learning rates](https://machinelearningmastery.com/understand-the-dynamics-of-learning-rate-on-deep-learning-neural-networks/#).